### PR TITLE
Fix hover with period alignment points and improve positioning of spikes and unified hover label

### LINKED ISSUE #5721: x-unified hover tooltips overflow
## Synopsis

The hover tooltips in x-unified mode exceed max-width and therefore overflow. This happens when:

- the chart is small and close to the sides 
- the chart is large (near width size) on a small screen - e.g. mobile
- the tooltip is displayed for a point on the chart that falls near the mid-point of the x-axis

The issue persists in Plotly v2.0.0.

## Example

You can see an example of the issue on our [live service](https://coronavirus.data.gov.uk/details/vaccinations#card-people_who_have_received_1st_dose_vaccinations_by_report_date), provided that your browser width is smaller than ~580px.

Screenshot 2021-06-08 at 14 52 22

## Expected behaviour

The hover tooltip x-extrema must always fall between x-extrema of the browser. An even better outcome would be for the tooltip to not spill out of the cartesian axes. 

I don't think there's an easy / immediate solution to this, primarily because the tooltip position moves based on cursor position. I suppose one way to address it is to do what folks at Google Charts have done - i.e. move the tooltip until it hits one side, then stop moving horizontally. 

Here's an example, which you've probably seen before:

google-sample
[Credit: Google]

Thanks again for all your efforts.

### LINKED ISSUE #5822: hover misplaced with period positioning
In this pen the hover labels positioning is quite broken: https://codepen.io/nicolaskruchten/pen/gOWgpEx?editors=0010

when I hover directly on some of the points it's correct, but in general it's too far to the left. This doesn't happen when I use just the first trace, and when I set `xperiod` to e.g. M1.

### LINKED ISSUE #5841: Bars and unified hover location with half-years periods
This pen has oddly-positioned hovers: https://codepen.io/nicolaskruchten/pen/RwVZgBW?editors=0010

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #5822, fixes #5841 and fixes #5721.

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
This looks good AFAICT! @alexcjohnson can we get an expedited review plz?

File: test/jasmine/tests/hover_label_test.js [hunk 1]
```diff
@@ -5377,36 +5377,134 @@ describe('hovermode: (x|y)unified', function() {
...
                 ]});
 
                 _hover(gd, { xpx: 100, ypx: 200 });
+                assertLabel({title: 'Jan 1, 2000', items: [
+                    'bar : 1'
+                ]});
+
+                _hover(gd, { xpx: 210, ypx: 200 });
                 assertLabel({title: 'Jan', items: [
                     'bar : (Jan 1, 2000, 1)',
```
Comment (alexcjohnson) [inline test/jasmine/tests/hover_label_test.js:5390]: Line [5348](https://github.com/plotly/plotly.js/pull/5846/files#diff-cfd23075dfa021e046f24be2b4f2cf2ee1a8e86aeec566e4d66dea1c09218140R5348) has a typo that's making this test confusing (`xhoverfrmat`)

Comment (archmoj) [reply test/jasmine/tests/hover_label_test.js]: Good eye. Addressed in https://github.com/plotly/plotly.js/pull/5846/commits/7c37e982f164cc6951954fbbb9c7adb46b5f5385.

**Comment by alexcjohnson** [comment]
Looking specifically at the test I [commented on above](https://github.com/plotly/plotly.js/pull/5846#discussion_r674135049) but having fixed the typo, here's the behavior on master:
pre-5846
It's not perfect, I don't think we should be showing Jan while the cursor is over the first half of Feb, but otherwise I actually think it looks correct: the spikeline shows the point that wins and the hover label shows all traces that match that period and it's positioned at one edge of the period.

Here's the behavior currently with this PR:
with-5846
Now the hoverlabel is positioned at the spikeline regardless of the period, it very often does NOT contain all the traces with points in a matching period, and when you do get multiple traces you mix points representing different periods - with some very weird flipping between two winning points right around the period boundary.

I thought at one point we had the idea that a period-positioned scatter trace should behave like a bar for hover purposes, in that any given point should be included in the hover whenever your cursor is over the period it represents. Did we drop that idea?

**Comment by nicolaskruchten** [comment]
> a period-positioned scatter trace should behave like a bar for hover purposes

Yes, here are the period-positioning invariants: https://github.com/plotly/plotly.js/issues/5553

The behaviour you note above is indeed troubling. Which mock is this?

**Comment by nicolaskruchten** [comment]
I'm also seeing some odd spikeline/hover mismatches in the all-bar case: https://codepen.io/nicolaskruchten/pen/bGWoVKB?editors=0010

unified

**Comment by alexcjohnson** [comment]
It's this, from the test I commented on:
```js
var scatterType = 'scatter';
Plotly.newPlot(gd, {
                data: [
                    {
                        name: 'bar',
                        type: 'bar',
                        x: ['2000-01', '2000-02'],
                        y: [1, 2],
                        xhoverformat: '%b',
                        xperiod: 'M1'
                    },
                    {
                        name: 'start',
                        type: scatterType,
                        x: ['2000-01', '2000-02'],
                        y: [1, 2],
                        xhoverformat: '%b',
                        xperiod: 'M1',
                        xperiodalignment: 'start'
                    },
                    {
                        name: 'end',
                        type: scatterType,
                        x: ['2000-01', '2000-02'],
                        y: [1, 2],
                        xhoverformat: '%b',
                        xperiod: 'M1',
                        xperiodalignment: 'end'
                    },
                ],
                layout: {
                    showlegend: false,
                    width: 600,
                    height: 400,
                    hovermode: 'x unified'
                }
            })
```

**Comment by alexcjohnson** [comment]
Better, it's now consistently showing all the traces I think it should. I still think we need to push the hover label outside the period the winning point represents, and right now it seems the hover label and spike line positions are referenced to different points: the label at the first (?) point in the hover set, but the spike line at the winning point. For example in the `period_positioning3` mock, with my cursor near the end of May, I can't see the brown point or most of the red point being hovered on, and the spike line is obscured:
Screen Shot 2021-07-21 at 4 56 21 PM

**Comment by nicolaskruchten** [comment]
Agreed with Alex's point above: the hoverlabel should always be to the side of the spikeline (as it is on master I believe)

**Comment by nicolaskruchten** [comment]
With this build, I don't seem to be able to hover on the left-most points in the figure from https://github.com/plotly/plotly.js/issues/5822

**Comment by archmoj** [comment]
> With this build, I don't seem to be able to hover on the left-most points in the figure from #5822 ... see https://codepen.io/nicolaskruchten/pen/MWmEzeJ?editors=0010

Good catch. Thank you!

**Comment by nicolaskruchten** [comment]
I think the change to this figure https://codepen.io/nicolaskruchten/pen/poPWqbL?editors=0010 is too large: I don't like that the hover-labels are now way over to the right of the bar. I think we should leave the hoverlabel right next to the spikeline like it is on master. That may occlude an "end" point in some cases when it is not the winning point, but we can live with that, and no one has complained about that yet.

**Comment by alexcjohnson** [comment]
> With this build, I don't seem to be able to hover on the left-most points in the figure from #5822

This is a tricky case: those left-most points refer to 1H 2013, a period which ends exactly at the left edge of the chart. So you can never have your cursor over that region, and thus if we ever show this hover label it will violate one of the invariants in #5553:

> a period-positioned point must not "win the hover" if the cursor is not within the period that the point represents

So do we relax that invariant? If so, in what principled way?

> I think the change to this figure https://codepen.io/nicolaskruchten/pen/poPWqbL?editors=0010 is too large

I actually quite like this: the spike line shows what point you're really focused on, but the extra space shows that you're also getting information about the whole bar. But I can see your argument, I'd be OK with pushing the label back to the spike line at least when it's a scatter point that wins. I think if it's a bar that wins it's still better to push it to the edge of the bar, and in fact to the edge of the bar group when we're labeling the whole group.

**Comment by nicolaskruchten** [comment]
> if we ever show this hover label it will violate one of the invariants

Zing! Ya got me :) OK, we should stick with the invariant and in cases like this the user will need to increase the range or rely on marker-padding or something.

**Comment by archmoj** [comment]
The https://github.com/plotly/plotly.js/pull/5846/commits/9716ef07e55e98383b6edd5228a58aa68d32a530 commit could also help address situation described in #5721.

**Comment by nicolaskruchten** [comment]
> But I can see your argument, I'd be OK with pushing the label back to the spike line at least when it's a scatter point that wins. I think if it's a bar that wins it's still better to push it to the edge of the bar, and in fact to the edge of the bar group when we're labeling the whole group.

@archmoj can we do this please? When the scatter point wins, it's ok for the hovelabel to occlude the bar.

**Comment by archmoj** [comment]
I can give it a try while you are reviewing & testing `(x|y) unified` label positions on various period & non-period mocks considering improvements made in `src/components/fx/hover.js` please : )

**Comment by archmoj** [comment]
> > But I can see your argument, I'd be OK with pushing the label back to the spike line at least when it's a scatter point that wins. I think if it's a bar that wins it's still better to push it to the edge of the bar, and in fact to the edge of the bar group when we're labeling the whole group.
> 
> @archmoj can we do this please? When the scatter point wins, it's ok for the hovelabel to occlude the bar.

Good idea. Applied in https://github.com/plotly/plotly.js/pull/5846/commits/16f4554b76f95956a40789795b5910764a0b3cfb.

**Comment by alexcjohnson** [comment]
It's still possible to have the hover label occluding the spikeline - hover on the right half of any period in `period_positioning3`
Screen Shot 2021-07-22 at 5 16 30 PM

**Comment by nicolaskruchten** [comment]
I think this is a case of the spike line not matching the hover-winning point, which is indeed a bug, and it seems worse with this PR than on master. Can we see if there's a fix for it plz?

**Comment by archmoj** [comment]
> It's still possible to have the hover label occluding the spikeline - hover on the right half of any period in `period_positioning3`
> Screen Shot 2021-07-22 at 5 16 30 PM

That's simply because of "When the scatter point wins, it's ok for the hovelabel to occlude the bar" logic added in https://github.com/plotly/plotly.js/pull/5846/commits/16f4554b76f95956a40789795b5910764a0b3cfb. Without it there would be no occlusion with spikeline :)

**Comment by archmoj** [comment]
> I think this is a case of the spike line not matching the hover-winning point, which is indeed a bug, and it seems worse with this PR than on master. Can we see if there's a fix for it plz?

This is also true : )

**Comment by nicolaskruchten** [comment]
Let's try to fix the spike line position too please before merging this. It should not be possible for the spike line to be on a different point that the winning point: there's a specific attribute for that.

**Comment by archmoj** [comment]
> Let's try to fix the spike line position too please before merging this. It should not be possible for the spike line to be on a different point that the winning point: there's a specific attribute for that.

On it.

**Comment by archmoj** [comment]
> Let's try to fix the spike line position too please before merging this. It should not be possible for the spike line to be on a different point that the winning point: there's a specific attribute for that.

Done in https://github.com/plotly/plotly.js/pull/5846/commits/dca5b1ec95dac51aad3e501d42fbd2384b5da279.

**Comment by nicolaskruchten** [comment]
Closer! So in this pen: https://codepen.io/nicolaskruchten/pen/xxdPKwZ?editors=0010 when I hover over the left-most point I get no bar in the hoverlabel, and when I hover on the next point I get the left-most bar in the hoverset. This is even more apparent when `hovermode = 'x'`.

I should add that the vertical position over the hoverlabels seems too low in this pen: https://codepen.io/nicolaskruchten/pen/QWvOLjJ?editors=0010 ... the hoverlabel is always around the bottom two points for me.

**Comment by archmoj** [comment]
> I should add that the vertical position over the hoverlabels seems too low in this pen: https://codepen.io/nicolaskruchten/pen/QWvOLjJ?editors=0010 ... the hoverlabel is always around the bottom two points for me.

Good catch. Fixed in https://github.com/plotly/plotly.js/pull/5846/commits/df6353d4e68c59fd013129e7a7f50a48f3fd7a3a

**Comment by archmoj** [comment]
> Closer! So in this pen: https://codepen.io/nicolaskruchten/pen/xxdPKwZ?editors=0010 when I hover over the left-most point I get no bar in the hoverlabel, and when I hover on the next point I get the left-most bar in the hoverset. This is even more apparent when `hovermode = 'x'`.

That's a bug.

**Comment by archmoj** [comment]
> > Closer! So in this pen: https://codepen.io/nicolaskruchten/pen/xxdPKwZ?editors=0010 when I hover over the left-most point I get no bar in the hoverlabel, and when I hover on the next point I get the left-most bar in the hoverset. This is even more apparent when `hovermode = 'x'`.
> 
> That's a bug.

Addressed in https://github.com/plotly/plotly.js/pull/5846/commits/9217db4e7023acf479cb23bd110ef4b593b680c4.

**Comment by nicolaskruchten** [comment]
OK, this last set of changes nails down everything I care about! There's still some oddity around end-positioned grouped-bars but we can leave that out for now. 

@alexcjohnson over to you, again :)

**Review by alexcjohnson — APPROVED** [review]
💃 I think we have a winner! 🏆

**Pull request merged**